src/libsyntax/parse/attr.rs

```diff
@@ -55,26 +76,49 @@ impl<'a> Parser<'a> {
     /// If permit_inner is true, then a leading `!` indicates an inner
     /// attribute
     pub fn parse_attribute(&mut self, permit_inner: bool) -> PResult<'a, ast::Attribute> {
-        debug!("parse_attributes: permit_inner={:?} self.token={:?}",
+        debug!("parse_attribute: permit_inner={:?} self.token={:?}",
                permit_inner,
                self.token);
+        let inner_parse_policy = if permit_inner {
+            InnerAttributeParsePolicy::Permitted
+        } else {
+            InnerAttributeParsePolicy::NotPermitted
+                { reason: DEFAULT_UNEXPECTED_INNER_ATTR_ERR_MSG }
+        };
+        self.parse_attribute_with_inner_parse_policy(inner_parse_policy)
+    }
+
+    /// The same as `parse_attribute`, except it takes in an `InnerAttributeParsePolicy`
+    /// that prescribes how to handle inner attributes.
+    fn parse_attribute_with_inner_parse_policy(&mut self,
+                                               inner_parse_policy: InnerAttributeParsePolicy)
+                                               -> PResult<'a, ast::Attribute> {
+        debug!("parse_attribute_with_inner_parse_policy: inner_parse_policy={:?} self.token={:?}",
+               inner_parse_policy,
+               self.token);
         let (span, value, mut style) = match self.token {
             token::Pound => {
                 let lo = self.span.lo;
                 self.bump();
 
-                if permit_inner {
+                if inner_parse_policy == InnerAttributeParsePolicy::Permitted {
                     self.expected_tokens.push(TokenType::Token(token::Not));
                 }
                 let style = if self.token == token::Not {
                     self.bump();
-                    if !permit_inner {
+                    if let InnerAttributeParsePolicy::NotPermitted { reason } = inner_parse_policy
+                    {
                         let span = self.span;
                         self.diagnostic()
-                            .struct_span_err(span,
-                                             "an inner attribute is not permitted in this context")
+                            .struct_span_err(span, reason)
                             .help("place inner attribute at the top of the module or \
                                    block")
+                            .note("an inner attribute is one of the form `#![foo]` in \
+                                   `mod bar { #![foo] }`, modifying the item enclosing it \
+                                   (`mod` bar in our example)")
+                            .note("an outer attribute is either of the form `#[foo]` in \
+                                   `#[foo] struct Baz` or is an outer doc comment, modifying \
+                                   the item following it (`struct Baz` in our example)")
```

Comment:
Yeah, I do agree that this is a lot of spew, unfortunately. But I also think the 'help' is pretty useful here since it is actually actionable, and especially so for a beginner.

Your suggestion for the note sounds pretty good to me, and I'm happy to change mine, but I have a vague worry that the sentence structure doesn't quite emphasize the "enclosing vs. following" distinction enough (when used as adjectives like that it somehow seems they don't stand out enough, if that makes any sense). Maybe it would be good to get the opinion of @steveklabnik on this? Also @Manishearth since he was the one who suggested the note.

Also, would you like it to be two separate notes or one long sentence/note?